library/ccm.c

```diff
@@ -99,7 +101,14 @@ int mbedtls_ccm_setkey( mbedtls_ccm_context *ctx,
         return( ret );
     }
 
-    return( ret );
+    if( keybits_dup == keybits && key_dup == key )
+    {
+        return( ret );
+    }
+
+    // In case of a FI - clear the context
+    mbedtls_ccm_free( ctx );
```

Comment:
@AndrzejKurek , why do we release the context? It was allocated bu user

Reply:
@AndrzejKurek it seems you wanted to release an internal context ctx->cipher_ctx

Reply:
I was considering releasing the whole context on failure (and, after a short investigation, it shouldn't be a problem), but fixed this one line. I commented on the rest of the changes below.

Reply:
@AndrzejKurek , we cannot release any buffer that was not allocated by us, in the same function. I wil give a couple of very problematic examples.
1. Application allocated context and will try to release it after it receive error value 
2. Application didn't allocate the context , it is global buffer or stack variable

Reply:
We should clean only the sensitive  data that was created in this function.

Reply:
How should we handle contexts that have sensitive data in them then?
For example, sha256 update() functions, or hmac_drbg seeding functions - if we detect a FI in them, the data in the context is potentially corrupted. What should we do in such situation?

Reply:
We should separate between corrupted data that we don't want our application to use  and between secretive data  that can leak. Zero key in ccm is not a problem , it will be zero if operation didn't succeed for any reason, not only FI, and in this case we should validate that error code will be returned and our application does not use this zero key. 
I think that we should clean all values we copied to ctx or output buffer if they are potentially secret, or can cause to wrong operation after it. As an example let's take https://github.com/ARMmbed/mbedtls/blob/2c16dfc05245d183a7567efe57f554b4c45aac35/library/hmac_drbg.c#L209
hmac_drbg_reseed_core function. One of the values that should be back to its previous value is reseed_counter. The context should not contain the information that the context was reseeded if it wasn't. By the way we should set it back in case of any FI attack (not only sensitive data).  I am not sure we should clean anything else in this specific function.    The argument buffer "additional" was copied to local buffer and there is no need to clean it.
Not related to my previous explanation -  In this specific function we have a bug, you don't release the context  if "additional" pointer was changed, you release it in case of other FI attack.

Reply:
OK, I'll remove context resetting from the reseed core function and other hmac functions.
What about SHA256 - would clearing the output for finish functions be enough? Should we, for the sha256_update_ret, remember the ctx->state from the beginning of the function and revert to it in case of FI?

Reply:
I think in sha256 it is enough to set 0 to ctx

Reply:
Do you mean setting the ctx->state to 0 in `mbedtls_internal_sha256_process`?
What about `mbedtls_sha256_update_ret`?

Reply:
I checked  mbedtls_sha256_free, it only set 0 to ctx, without release any memory. It is Okey to use it (as you did) or to set 0 to ctx in all required places.